[2.5 beta] circular references in tsconfig.extends result hanging devserver

### Describe the bug

Using tsconfig.extends it is possible to create circular references. 

tsconfig.json
```js
{
  "extends": "./tsconfig.circle.json"
  // ...
}
```
tsconfig.circle.json
```js
{
  "extends": "./tsconfig.json"
  // ...
}
```

The code that was addded in https://github.com/vitejs/vite/pull/4279 does not check this and runs forever, stalling the devserver.

note, the circular reference example is very short, it could be a lot less obvious

### Reproduction

```
npm init vite -t vanilla-ts ts-circle
npm i
# add circular reference tsconfig as shown above
npm run dev
# open devserver, it works bc tsconfig isn't read
npm run build
# fails, tsc correctly reports the circle
npm i -D vite@beta
# open devserver, it hangs, no errors on cli or browser console
```

### System Info

```shell
System:
    OS: Linux 5.12 Ubuntu 20.04.2 LTS (Focal Fossa)
    CPU: (12) x64 AMD Ryzen 5 4600H with Radeon Graphics
    Memory: 6.86 GB / 15.07 GB
    Container: Yes
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 14.17.1 - ~/.nvm/versions/node/v14.17.1/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v14.17.1/bin/yarn
    npm: 6.14.13 - ~/.nvm/versions/node/v14.17.1/bin/npm
  Browsers:
    Chrome: 92.0.4515.107
    Firefox: 90.0.2
  npmPackages:
    vite: ^2.5.0-beta.1 => 2.5.0-beta.1
```


### Used Package Manager

npm

### Logs

```shell
> tsc && vite build

error TS18000: Circularity detected while resolving configuration: /home/dominikg/develop/reproductions/vite-ts-vanilla/tsconfig.json -> /home/dominikg/develop/reproductions/vite-ts-vanilla/tsconfig.circle.json -> /home/dominikg/develop/reproductions/vite-ts-vanilla/tsconfig.json
```


### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/playground/tsconfig-json/nested/tsconfig.json
+++ b/packages/playground/tsconfig-json/nested/tsconfig.json
@@ -1,3 +1,4 @@
+// prettier-ignore
 {
   "include": ["./"],
   "compilerOptions": {

--- a/packages/playground/tsconfig-json/nested/tsconfig.json
+++ b/packages/playground/tsconfig-json/nested/tsconfig.json
@@ -14,6 +14,7 @@
     "noUnusedParameters": true,
     "noImplicitReturns": true,
 
+    /* tsconfig.json should support comments and trailing comma */
     "useDefineForClassFields": false
   }
 }

--- a/packages/playground/tsconfig-json/nested/tsconfig.json
+++ b/packages/playground/tsconfig-json/nested/tsconfig.json
@@ -15,6 +15,6 @@
     "noImplicitReturns": true,
 
     /* tsconfig.json should support comments and trailing comma */
-    "useDefineForClassFields": false
+    "useDefineForClassFields": false,
   }
 }

--- a/packages/vite/package.json
+++ b/packages/vite/package.json
@@ -117,6 +117,8 @@
     "source-map": "^0.6.1",
     "source-map-support": "^0.5.19",
     "strip-ansi": "^6.0.0",
+    "strip-bom": "^4.0.0",
+    "strip-json-comments": "^3.1.1",
     "terser": "^5.7.1",
     "tsconfig": "^7.0.0",
     "tslib": "^2.3.0",

--- a/packages/vite/src/node/plugins/esbuild.ts
+++ b/packages/vite/src/node/plugins/esbuild.ts
@@ -1,3 +1,4 @@
+import fs from 'fs'
 import path from 'path'
 import chalk from 'chalk'
 import { Plugin } from '../plugin'
@@ -14,7 +15,8 @@ import { SourceMap } from 'rollup'
 import { ResolvedConfig } from '..'
 import { createFilter } from '@rollup/pluginutils'
 import { combineSourcemaps } from '../utils'
-import { find as findTSConfig, readFile as readTSConfig } from 'tsconfig'
+import stripBom from 'strip-bom'
+import stripComments from 'strip-json-comments'
 import { createRequire } from 'module'
 
 const debug = createDebugger('vite:esbuild')
@@ -41,43 +43,6 @@ type TSConfigJSON = {
   [key: string]: any
 }
 type TSCompilerOptions = NonNullable<TSConfigJSON['compilerOptions']>
-
-const tsconfigCache = new Map<string, TSConfigJSON>()
-async function loadTsconfigJsonForFile(
-  filename: string
-): Promise<TSConfigJSON> {
-  const directory = path.dirname(filename)
-
-  const cached = tsconfigCache.get(directory)
-  if (cached) {
-    return cached
-  }
-
-  let configPath = await findTSConfig(directory)
-  let tsconfig: TSConfigJSON = {}
-
-  if (configPath) {
-    tsconfig = (await readTSConfig(configPath)) as TSConfigJSON
-    while (tsconfig.extends) {
-      const configRequire = createRequire(configPath)
-
-      const extendsPath = configRequire.resolve(tsconfig.extends)
-      const extendedConfig = (await readTSConfig(extendsPath)) as TSConfigJSON
-
-      tsconfig = {
-        extends: extendedConfig.extends,
-        compilerOptions: {
-          ...extendedConfig.compilerOptions,
-          ...tsconfig.compilerOptions
-        }
-      }
-      configPath = extendsPath
-    }
-  }
-
-  tsconfigCache.set(directory, tsconfig)
-  return tsconfig
-}
 
 export async function transformWithEsbuild(
   code: string,
@@ -233,4 +198,196 @@ function prettifyMessage(m: Message, code: string): string {
     res += `\n` + generateCodeFrame(code, offset, offset + 1)
   }
   return res + `\n`
+}
+
+// modified from <https://github.com/TypeStrong/tsconfig/blob/v7.0.0/src/tsconfig.ts#L75-L95>
+
+/**
+ * Copyright (c) 2015 TypeStrong
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+async function findTSConfig(dir: string): Promise<string | void> {
+  const configFile = path.resolve(dir, 'tsconfig.json')
+
+  return stat(configFile).then((stats) => {
+    if (isFile(stats)) {
+      return configFile
+    }
+
+    const parentDir = path.dirname(dir)
+
+    if (dir === parentDir) {
+      return
+    }
+
+    return findTSConfig(parentDir)
+  })
+}
+
+/**
+ * Check if a file exists.
+ */
+function stat(filename: string): Promise<fs.Stats | void> {
+  return new Promise<fs.Stats | void>((resolve, reject) => {
+    fs.stat(filename, (err, stats) => {
+      return err ? resolve(undefined) : resolve(stats)
+    })
+  })
+}
+
+/**
+ * Check filesystem stat is a directory.
+ */
+function isFile(stats: fs.Stats | void) {
+  return stats
+    ? (stats as fs.Stats).isFile() || (stats as fs.Stats).isFIFO()
+    : false
+}
+
+// from <https://github.com/TypeStrong/tsconfig/pull/31>
+// by @dominikg
+
+/**
+ * replace dangling commas from pseudo-json string with single space
+ *
+ * limitations:
+ * - pseudo-json must not contain comments, use strip-json-comments before
+ * - only a single dangling comma before } or ] is removed
+ *   stripDanglingComma('[1,2,]') === '[1,2 ]
+ *   stripDanglingComma('[1,2,,]') === '[1,2, ]
+ *
+ * implementation heavily inspired by strip-json-comments
+ */
+function stripDanglingComma(jsonString: string) {
+  /**
+   * Check if char at qoutePosition is escaped by an odd number of backslashes preceding it
+   */
+  function isEscaped(jsonString: string, quotePosition: number) {
+    let index = quotePosition - 1
+    let backslashCount = 0
+
+    while (jsonString[index] === '\\') {
+      index -= 1
+      backslashCount += 1
+    }
+
+    return Boolean(backslashCount % 2)
+  }
+  let insideString = false
+  let offset = 0
+  let result = ''
+  let danglingCommaPos = null
+  for (let i = 0; i < jsonString.length; i++) {
+    const currentCharacter = jsonString[i]
+
+    if (currentCharacter === '"') {
+      const escaped = isEscaped(jsonString, i)
+      if (!escaped) {
+        insideString = !insideString
+      }
+    }
+
+    if (insideString) {
+      danglingCommaPos = null
+      continue
+    }
+    if (currentCharacter === ',') {
+      danglingCommaPos = i
+      continue
+    }
+    if (danglingCommaPos) {
+      if (currentCharacter === '}' || currentCharacter === ']') {
+        result += jsonString.slice(offset, danglingCommaPos) + ' '
+        offset = danglingCommaPos + 1
+        danglingCommaPos = null
+      } else if (!currentCharacter.match(/\s/)) {
+        danglingCommaPos = null
+      }
+    }
+  }
+
+  return result + jsonString.substring(offset)
+}
+
+async function readTSConfig(configPath: string): Promise<TSConfigJSON> {
+  const content: string = await new Promise((resolve, reject) => {
+    fs.readFile(configPath, 'utf-8', (err, data) => {
+      if (err) {
+        reject(err)
+        return
+      }
+      resolve(stripComments(stripBom(data)))
+    })
+  })
+
+  // tsconfig.json can be empty
+  if (/^\s*$/.test(content)) {
+    return {}
+  }
+
+  return JSON.parse(stripDanglingComma(content)) as TSConfigJSON
+}
+
+const tsconfigCache = new Map<string, TSConfigJSON>()
+async function loadTsconfigJsonForFile(
+  filename: string
+): Promise<TSConfigJSON> {
+  const directory = path.dirname(filename)
+
+  const cached = tsconfigCache.get(directory)
+  if (cached) {
+    return cached
+  }
+
+  let configPath = await findTSConfig(directory)
+  let tsconfig: TSConfigJSON = {}
+
+  if (configPath) {
+    const visited = new Set()
+    visited.add(configPath)
+
+    tsconfig = await readTSConfig(configPath)
+    while (tsconfig.extends) {
+      const configRequire = createRequire(configPath)
+
+      const extendsPath = configRequire.resolve(tsconfig.extends)
+      const extendedConfig = await readTSConfig(extendsPath)
+
+      if (visited.has(extendsPath)) {
+        throw new Error(
+          `Circular dependency detected in the "extends" field of ${configPath}`
+        )
+      }
+      visited.add(extendsPath)
+
+      tsconfig = {
+        extends: extendedConfig.extends,
+        compilerOptions: {
+          ...extendedConfig.compilerOptions,
+          ...tsconfig.compilerOptions
+        }
+      }
+      configPath = extendsPath
+    }
+  }
+
+  tsconfigCache.set(directory, tsconfig)
+  return tsconfig
 }

--- a/packages/vite/src/node/plugins/esbuild.ts
+++ b/packages/vite/src/node/plugins/esbuild.ts
@@ -226,19 +226,18 @@ function prettifyMessage(m: Message, code: string): string {
 async function findTSConfig(dir: string): Promise<string | void> {
   const configFile = path.resolve(dir, 'tsconfig.json')
 
-  return stat(configFile).then((stats) => {
-    if (isFile(stats)) {
-      return configFile
-    }
+  const stats = await stat(configFile)
+  if (isFile(stats)) {
+    return configFile
+  }
 
-    const parentDir = path.dirname(dir)
+  const parentDir = path.dirname(dir)
 
-    if (dir === parentDir) {
-      return
-    }
+  if (dir === parentDir) {
+    return
+  }
 
-    return findTSConfig(parentDir)
-  })
+  return findTSConfig(parentDir)
 }
 
 /**

--- a/packages/vite/src/node/plugins/esbuild.ts
+++ b/packages/vite/src/node/plugins/esbuild.ts
@@ -245,9 +245,9 @@ async function findTSConfig(dir: string): Promise<string | void> {
  * Check if a file exists.
  */
 function stat(filename: string): Promise<fs.Stats | void> {
-  return new Promise<fs.Stats | void>((resolve, reject) => {
+  return new Promise((resolve, reject) => {
     fs.stat(filename, (err, stats) => {
-      return err ? resolve(undefined) : resolve(stats)
+      return err ? resolve() : resolve(stats)
     })
   })
 }
@@ -256,9 +256,7 @@ function stat(filename: string): Promise<fs.Stats | void> {
  * Check filesystem stat is a directory.
  */
 function isFile(stats: fs.Stats | void) {
-  return stats
-    ? (stats as fs.Stats).isFile() || (stats as fs.Stats).isFIFO()
-    : false
+  return stats ? stats.isFile() || stats.isFIFO() : false
 }
 
 // from <https://github.com/TypeStrong/tsconfig/pull/31>
@@ -288,8 +286,9 @@ function stripDanglingComma(jsonString: string) {
       backslashCount += 1
     }
 
-    return Boolean(backslashCount % 2)
+    return backslashCount % 2 === 1
   }
+
   let insideString = false
   let offset = 0
   let result = ''

